felix/calc/active_bpgpeer_calculator.go

```diff
@@ -0,0 +1,299 @@
+		if selector.Evaluate(labels) {
+			return EpPeerData{
+				v3PeerName: name,
+			}
+		}
+	}
+
+	// Otherwise return an empty peer data.
+	return EpPeerData{}
+}
+
+// Given a new peer data, check and update the cache if needed.
+func (abp *ActiveBGPPeerCalculator) checkAndUpdatePeerData(id model.WorkloadEndpointKey, newPeerData EpPeerData) {
+	// Get current peer data of the endpoint.
+	peerData := abp.peersByID[id]
+
+	// If peer data hasn't changed, we don't need to send any update.
+	if newPeerData == peerData {
+		return
+	}
+
+	// Update peer data for the endpoint.
+	if newPeerData.isEmpty() {
+		delete(abp.peersByID, id)
+	} else {
+		abp.peersByID[id] = newPeerData
+	}
+
+	// Send the update.
+	logrus.Debugf("Send BGP Peer data update %s on endpoint %s", newPeerData.v3PeerName, id)
+	abp.OnEndpointBGPPeerDataUpdate(id, &newPeerData)
+}
+
+func (abp *ActiveBGPPeerCalculator) onEndpointUpdate(id model.WorkloadEndpointKey, newLabels map[string]string) {
+	// Save new label for this endpoint if it is different with the old one.
+	labels, exists := abp.labelsByID[id]
+	if exists {
+		if reflect.DeepEqual(labels, newLabels) {
+			return
+		}
+	}
+	abp.labelsByID[id] = newLabels
+
+	// Calculate peer data based on the new label.
+	newPeerData := abp.calculatePeerDataFromLabels(newLabels)
+
+	abp.checkAndUpdatePeerData(id, newPeerData)
+}
+
+func (abp *ActiveBGPPeerCalculator) onEndpointDelete(id model.WorkloadEndpointKey) {
+	// Find and delete the data for this endpoint.
+	if _, exists := abp.labelsByID[id]; !exists {
+		return
+	}
+	delete(abp.labelsByID, id)
+
+	// Get current peer data of the endpoint.
+	if _, exists := abp.peersByID[id]; !exists {
+		return
+	}
+
+	// Update peer data for the endpoint.
+	delete(abp.peersByID, id)
+
+	// Send the update.
+	logrus.Debugf("Clear BGP Peer data on endpoint %s", id)
+	abp.OnEndpointBGPPeerDataUpdate(id, nil)
+}
+
+func (abp *ActiveBGPPeerCalculator) ifBgpPeerSelectHost(bgpPeer *v3.BGPPeer) bool {
+	if bgpPeer.Spec.Node == abp.hostname {
+		return true
+	}
+
+	selector, err := sel.Parse(bgpPeer.Spec.NodeSelector)
+	if err != nil {
+		logrus.WithError(err).Errorf("Couldn't parse selector: %s", bgpPeer.Spec.NodeSelector)
+		return false
+	}
+
+	if len(abp.nodeLabels) != 0 {
+		if selector.Evaluate(abp.nodeLabels) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (abp *ActiveBGPPeerCalculator) onBGPPeerUpdate(bgpPeer *v3.BGPPeer) {
+	logrus.WithField("bgppeer", bgpPeer).Debugf("On BGP Peer update")
+	name := bgpPeer.Name
+
+	abp.bgpPeersByName[name] = bgpPeer
+
+	rawSelector := bgpPeer.Spec.LocalWorkloadSelector
+	newSelector, err := sel.Parse(rawSelector)
+	if err != nil {
+		logrus.WithError(err).Errorf("Couldn't parse selector: %s", rawSelector)
+		return
+	}
+
+	// Save new selector for this BGP peer if it is different with the old one.
+	selector, exists := abp.workloadSelectorsByPeerName[name]
+	if exists {
+		if reflect.DeepEqual(selector, newSelector) {
```

Comment:
selector has cached fields inside so DeepEqual is incorrect.  I think it has an Equal method?  (Or you can compare the UID)